File: src/1_100/0082_Remove_Duplicates_from_Sorted_List_II/Remove_Duplicates_from_Sorted_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
    	if(head == null) {
    		return head;
    	}
    	ListNode dummy = new ListNode(-1);
    	ListNode low = head;
    	ListNode fast = head.next;
    	dummy.next = head;
    	ListNode p = dummy;
    	while(fast!=null && fast.next!=null) {
    		if(low.val != fast.val) {
    			p.next = low;
    			p = p.next;
    			low = fast;
    			fast = fast.next;
    		} else {
    			while(fast.next!=null && fast.next.val==low.val) {
    				fast = fast.next;
    			}
    			low = fast.next;
    			fast = fast.next!=null? fast.next.next : null;
    		}
    	}
    	if(fast != null) {
    		if(low.val == fast.val) {
    			p.next = null;
    		}else {
    			p.next = low;
    			p.next.next = fast;
    		}
    	} else {
    		p.next = low!=null? low : null;
    	}
    	return dummy.next;
    }
}

File: src/1_100/0082_Remove_Duplicates_from_Sorted_List_II/Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not (head and head.next):
            return head
        p = ListNode(-1)
        low,fast,p.next = p,head,head
        while fast and fast.next:
            if low.next.val!=fast.next.val:
                low,fast = low.next,fast.next
            else:
                while fast.next and low.next.val==fast.next.val:
                    fast = fast.next
                fast,low.next = fast.next,fast.next
        return p.next
		
		
		
	def deleteDuplicates(self, head):
		p = ListNode(-1)
		a,b,p.next = p,head,head
		while b and b.next:
			if a.next.val!=b.next.val:
				a,b = a.next,b.next
			else:
				while b.next and a.next.val==b.next.val:
					b = b.next
				a.next,b = b.next,b.next
		return p.next
		
		